test/rubygems/test_gem_safe_marshal.rb

```diff
@@ -0,0 +1,336 @@
+      assert_safe_load_as(h)
+    end
+  end
+
+  def test_hash_with_default_value
+    assert_safe_load_as Hash.new([])
+  end
+
+  def test_hash_with_compare_by_identity
+    pend "`read_user_class` not yet implemented"
+
+    assert_safe_load_as Hash.new.compare_by_identity
+  end
+
+  def test_frozen_object
+    assert_safe_load_as Gem::Version.new("1.abc").freeze
+  end
+
+  def test_date
+    assert_safe_load_as Date.new(1994, 12, 9)
+  end
+
+  def test_rational
+    pend "truffleruby dumps rationals with ivars set on array, see https://github.com/oracle/truffleruby/issues/3228" if RUBY_ENGINE == "truffleruby"
+
+    assert_safe_load_as Rational(1, 3)
+  end
+
+  [
+    0, 1, 2, 3, 4, 5, 6, 122, 123, 124, 127, 128, 255, 256, 257,
+    2**16, 2**16 - 1, 2**20 - 1,
+    2**28, 2**28 - 1,
+    2**32, 2**32 - 1,
+    2**63, 2**63 - 1
+  ].
+  each do |i|
+    define_method("test_int_ #{i}") do
+      assert_safe_load_as i
+      assert_safe_load_as(-i)
+      assert_safe_load_as(i + 1)
+      assert_safe_load_as(i - 1)
+    end
+  end
+
+  def test_gem_spec_disallowed_symbol
+    e = assert_raise(Gem::SafeMarshal::Visitors::ToRuby::UnpermittedSymbolError) do
+      spec = Gem::Specification.new do |s|
+        s.name = "hi"
+        s.version = "1.2.3"
+
+        s.dependencies << Gem::Dependency.new("rspec", Gem::Requirement.new([">= 1.2.3"]), :runtime).tap {|d| d.instance_variable_set(:@name, :rspec) }
+      end
+      Gem::SafeMarshal.safe_load(Marshal.dump(spec))
+    end
+
+    assert_equal e.message, "Attempting to load unpermitted symbol \"rspec\" @ root.[9].[0].@name"
+  end
+
+  def assert_safe_load_marshal(dumped, additional_methods: [], permitted_ivars: nil, equality: true, marshal_dump_equality: true)
+    loaded = Marshal.load(dumped)
+    safe_loaded =
+      if permitted_ivars
+        with_const(Gem::SafeMarshal, :PERMITTED_IVARS, permitted_ivars) do
+          Gem::SafeMarshal.safe_load(dumped)
+        end
+      else
+        Gem::SafeMarshal.safe_load(dumped)
+      end
+
+    # NaN != NaN, for example
+    if equality
+      assert_equal loaded, safe_loaded, "should equal what Marshal.load returns"
+    end
+
+    assert_equal loaded.to_s, safe_loaded.to_s, "should have equal to_s"
+    assert_equal loaded.inspect, safe_loaded.inspect, "should have equal inspect"
+    additional_methods.each do |m|
+      if m.is_a?(Proc)
+        call = m
+      else
+        call = ->(obj) { obj.__send__(m) }
+      end
+
+      assert_equal call[loaded], call[safe_loaded], "should have equal #{m}"
+    end
+    if marshal_dump_equality
+      assert_equal Marshal.dump(loaded).dump, Marshal.dump(safe_loaded).dump, "should Marshal.dump the same"
+    end
+  end
+
+  def assert_safe_load_as(x, **kwargs)
+    dumped = Marshal.dump(x)
+    equality = x == x # rubocop:disable Lint/BinaryOperatorWithIdenticalOperands
+    assert_safe_load_marshal(dumped, equality: equality, **kwargs)
+  end
+
+  def with_const(mod, name, new_value, &block)
```

Comment:
This is used to temporarily modify `PERMITTED_IVARS` of `Gem::SafeMarshal`, since that is used in `safe_load` method. Wouldn't be easier using `Gem::SafeMarshal.load` with `permitted_ivars` attribute for such a purpose directly?